### ISSUE TITLE
run_e2e_workflow.py should support using python functions to generate the workflow

### ISSUE THREAD

**Issue by jlewi** [kubeflow/testing#423]
Currently we use ksonnet to parameterize our Argo workflows in our E2E tests.
Since ksonnet is being deprecated we need a different way to parameterize our E2E tests.
I think a flexible way to do this is to let people define a class in python with methods that return one or more Argo workflows.

```
class MyTestWorkflows(object):
   def workflow1(self, param1=1)
        ...
        return workflow
....
```

Our prow_config.yaml files could then look like the following

```
workflows
- py_func: package.module.class.method_name
   kw_args:
       arg1: value
```

Then run_e2e_workflow.py can 
   1. Load prow_config.yaml
   1. Invoke py_func with the specified arguments
   1. submit and wait for the supplied Argo workflow

just invoke `MyTestWorkflows` to get the workflow.

When we make this change in the same PR we should replace the existing ksonnet test in kubeflow/testing. This way the presubmit will verify that the new code is working.

Related to: kubeflow/kubeflow#3035 migrate off ksonnet

@scottilee Any interest in picking this up?

**Comment by scottilee** [comment]
@jlewi Sure, I can give it a try.

/assign

**Comment by scottilee** [comment]
@jlewi Just started looking at this today and need some more context and make sure this is the correct path ahead:
1) Edit `parse_config_file` function to parse a `py_func` field along with its associated args
2) Edit the `run` function to remove the `ks_cmd`s and instead invoke the python method -- is the latter as simple as updating the `util.run` calls to execute the python methods?

Also, is there a way I can test my work in some kind of local setup to make sure it's working correctly?

**Comment by jlewi** [comment]
I think we want to continue to support both ksonnet and py_func. 

In the loop over workflows here
https://github.com/kubeflow/testing/blob/1231ff98ec13b483acc4eb17c158f25e7e5b1595/py/kubeflow/testing/run_e2e_workflow.py#L178

We want to support 2 types of workflows

1. Ksonnet - which would use existing code
2. py_func

In the case of py_func you'll want to invoke the func to get a json description describing the workflow.

Then you'll want to create that workflow by invoking argo_client.

run_e2e_workflow.py has a unittest
https://github.com/kubeflow/testing/blob/master/py/kubeflow/tests/run_e2e_workflow_test.py

So I would suggest updating that unittest to verify your new code is working.

If you create a pull request then your presubmit should end up using your modified run_e2e_workflow.py.

So in the PR you create you should update 
https://github.com/kubeflow/testing/blob/master/prow_config.yaml

To include a new py_func test so that you can use it to verify things are working.

**Comment by kkasravi** [comment]
@scottilee I'm defining a run_e2e_workflow.py in the new [kfctl repo](https://github.com/kubeflow/kfctl). If you open a PR with the py_func definition I would be interested in reviewing what you have and try using it in the new kfctl repo.  I've done a bit of experimentation with class decorators but would be interested in using what you come up with.
thanks Kam

**Comment by jlewi** [comment]
@kkasravi Why would run_e2e_workflow.py live in kubeflow/kfctl? Don't we want a single script that is used across repositories?

/cc @krishnadurai @swiftdiaries

**Comment by kkasravi** [comment]
@jlewi 
yes - sorry, the workflows are in kfctl/prow_config.yaml. The py_func definitions are in kfctl.

**Comment by swiftdiaries** [comment]
Yeah currently the design is to manually copy over the Workflow components if we want to re-use existing tests across repos.
See example here: https://github.com/kubeflow/manifests/blob/master/tests/workflows/components/kfctl_go_test.jsonnet
This adds extra maintenance overhead, going forward like Jeremy mentioned we should have a single script that can be re-used.

**Comment by jlewi** [comment]
#431 is being reverted in #449. It looks like #431 is always adding workflows to be waited on even if the workflow is skipped. We will need to roll #431 forward and fix that issue.

@scottilee Do you want to take a look? 

/cc @kkasravi

**Comment by kkasravi** [comment]
@jlewi @scottilee 

It looks like some of the code is repeated when the workflow is 'app_dir'. (create_started_file)
@scottilee did you want to have a go at fixing it?
If not, I can make a fix and push a PR. @scottilee if you want to merge my fix into #431 that would work too. Let me know. I would encourage you to try.

If you wanted me to go forward with a fix, I'll do the following
1. checkout #431 
2. rebase from master
3. make the fix, push a new PR
4. @scottilee you can update #431 with the changes and we can roll that forward
5. or we can roll the PR I push forward.

**Comment by scottilee** [comment]
I think I found the issue. I put up another PR here: https://github.com/kubeflow/testing/pull/451
